File: src/main/java/ru/citeck/ecos/history/converter/impl/HistoryRecordConverter.java
package ru.citeck.ecos.history.converter.impl;

import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Service;
import ru.citeck.ecos.history.converter.Converter;
import ru.citeck.ecos.history.domain.HistoryRecordEntity;
import ru.citeck.ecos.history.dto.HistoryRecordDto;

import java.lang.reflect.Field;
import java.time.ZoneId;
import java.util.*;

@Slf4j
@Service("historyRecordConverter")
public class HistoryRecordConverter implements Converter<HistoryRecordEntity, HistoryRecordDto> {

    @Override
    public HistoryRecordDto convert(HistoryRecordEntity historyRecordEntity) {
        HistoryRecordDto result = new HistoryRecordDto();

        Long id = historyRecordEntity.getId();
        result.setUuid(id != null ? id.toString() : "no_uuid");

        result.setHistoryEventId(historyRecordEntity.getHistoryEventId());
        result.setUserId(historyRecordEntity.getUserId());
        result.setUsername(historyRecordEntity.getUsername());
        result.setComments(historyRecordEntity.getComments());
        result.setVersion(historyRecordEntity.getVersion());
        result.setEventType(historyRecordEntity.getEventType());

        Date creationTime = historyRecordEntity.getCreationTime();
        result.setCreationTime(creationTime != null ? creationTime.getTime() : 0L);

        result.setTaskTitle(historyRecordEntity.getTaskTitle());
        result.setTaskRole(historyRecordEntity.getTaskRole());
        result.setTaskOutcome(historyRecordEntity.getTaskOutcome());
        result.setTaskOutcomeName(historyRecordEntity.getTaskOutcomeName());
        result.setTaskDefinitionKey(historyRecordEntity.getTaskDefinitionKey());
        result.setTaskType(historyRecordEntity.getFullTaskType());
        result.setDocumentId(historyRecordEntity.getDocumentId());
        return result;
    }

    public Map<String, String> toMap(HistoryRecordDto historyRecordDto){
        HashMap<String, String> propertyMap = new HashMap<>();
        Field[] fields = historyRecordDto.getClass().getDeclaredFields();
        for (Field property : fields) {
            property.setAccessible(true);
            Object objectValue = null;
            try {
                objectValue = property.get(historyRecordDto);
                propertyMap.put(property.getName(), objectValue!=null? String.valueOf(objectValue): null);
            } catch (IllegalArgumentException | IllegalAccessException exception) {
                log.error(String.format("Failed to get attribute '%s' for %s", property.getName(), historyRecordDto),
                    exception);
            }
        }
        return propertyMap;
    }
}

File: src/main/java/ru/citeck/ecos/history/converter/HistoryRecordConverter.kt
package ru.citeck.ecos.history.converter

import org.springframework.stereotype.Component
import ru.citeck.ecos.commons.json.Json
import ru.citeck.ecos.history.domain.HistoryRecordEntity
import ru.citeck.ecos.history.dto.HistoryRecordDto

@Component
class HistoryRecordConverter {

    fun toDto(entity: HistoryRecordEntity): HistoryRecordDto {

        val result = HistoryRecordDto()

        result.uuid = entity.id?.toString() ?: "no_uuid"
        result.historyEventId = entity.historyEventId
        result.userId = entity.userId
        result.username = entity.username
        result.comments = entity.comments
        result.version = entity.version
        result.eventType = entity.eventType
        result.creationTime = entity.creationTime?.toEpochMilli() ?: 0L
        result.taskTitle = entity.taskTitle
        result.taskRole = entity.taskRole
        result.taskOutcome = entity.taskOutcome
        result.taskOutcomeName = entity.taskOutcomeName
        result.taskDefinitionKey = entity.taskDefinitionKey
        result.taskType = entity.fullTaskType
        result.taskCompletedOnBehalfOf = entity.taskCompletedOnBehalfOf
        result.documentId = entity.documentId

        return result
    }

    fun toDto(sources: List<HistoryRecordEntity>): List<HistoryRecordDto> {
        return sources.map { toDto(it) }
    }

    fun toMap(dto: HistoryRecordDto): Map<String, String> {
        val mapType = Json.mapper.getMapType(String::class.java, String::class.java)
        val resultWithNull = Json.mapper.convert<Map<String, String?>>(dto, mapType)
            ?: error("Error while dto-to-map conversion. Dto: $dto")
        val result = HashMap<String, String>()
        for ((k, v) in resultWithNull) {
            if (v != null) {
                result[k] = v
            }
        }
        return result
    }
}
